### ISSUE TITLE
application should block adding a bare repository

### ISSUE THREAD

**Issue by ysabri** [desktop/desktop#4293]
## Description
So when I try and create a new repository in a directory that includes a bare repository, the app recognizes that the directory includes a git repository and asks me to add it instead. So far so good until I add the repository and the app tries to perform a git operation (I don't know which one exactly) and fails with "fatal: This operation must be run in a work tree". 

## Version

* GitHub Desktop: 1.1.0

* Operating system: Windows 10

## Steps to Reproduce

1. Add a bare repository.  
### Expected Behavior

The app rejects adding it since it does not have a working tree or offers to do a local clone.
### Actual Behavior

The app adds it and fails to do a git operation repeatedly till the repo is removed. 

## Additional Information

So I looked at how the app decides if a path is a git directory. I understand how rev-parse --show-cdup is useful but it does not care if the repo is bare or not. I also tried to put a path into the url field of clone and the app complained that the destination already exists. I feel local clones is an enhancement that should be added as well.

**Comment by tierninho** [comment]
@ysabri thanks for filing. Possible to include the logs for this before one of our core members addresses your issue? See the Help menu to find the logs in the Explorer.

Have you tried running one of the following?

Single commands:
- `git --work-tree=/path/to/work/tree checkout master`

Environment variable
- `GIT_WORK_TREE=/path/to/work/tree git status`

**Comment by ysabri** [comment]
Here is the file: [2018-03-20.desktop.production.log](https://github.com/desktop/desktop/files/1831153/2018-03-20.desktop.production.log)
 I also I am not sure what path to feed --work-tree, I do not know what in the bare repository to point the command. I tried refs/heads and it did not work.

**Comment by shiftkey** [comment]
> So when I try and create a new repository in a directory that includes a bare repository, the app recognizes that the directory includes a git repository and asks me to add it instead.

To confirm I understand the issue, are you using `git init --bare` to create a Git repository and then adding that folder to the app?

```
$ cd test
$ mkdir repo
$ cd repo
$ git init --bare
Initialized empty Git repository in C:/Users/shiftkey/src/test/repo/
$ ls
config  description  HEAD  hooks/  info/  objects/  refs/
```

I suspect there are some assumptions in the app around the presence of a `.git` directory _within_ the folder that you provide. @ysabri can you tell me more about your use case here, and why you're avoiding initialising the repository in the typical way?

**Comment by ysabri** [comment]
Sorry I approached the question weirdly. I have a bare repo with content in it already, meaning if I clone it I will be able to access a worktree with several commits in it. I tried to add it to the app and the add was successful, meaning the app recognizes it is a repository but _not that it is a bare repository_.  My suggestion is for the app to able to recognize a bare repo, prevent me from adding it and maybe suggest that I perform a local clone of it. 

I am agreeing with the fact that the app prevented me from creating a new repository where the bare repository resides. What I am not agreeing with is that the app let me add it. This stems from the fact that rev-prase --show-cdup does not convey information about the repository being bare or not, which might lead to the false assumption that a bare repository is addable to the app.

I know this a weird case, but it is not impossible, I am just pointing it out. I know how to setup the repository normally and use it, I was just trying it out on the app when I saw the app only uses --show-cdup option.

**Comment by batarmanrookie** [comment]
Because  at begening i  was hack by this fucking shift and im not the best in english and i learn all by myself 😳👻💩lol

Envoyémy de mon iPhone

Le 20 mars 2018 à 19:19, Brendan Forster <notifications@github.com<mailto:notifications@github.com>> a écrit :


So when I try and create a new repository in a directory that includes a bare repository, the app recognizes that the directory includes a git repository and asks me to add it instead.

To confirm I understand the issue, are you using git init --bare to create a Git repository and then adding that folder to the app?

$ cd test
$ mkdir repo
$ cd repo
$ git init --bare
Initialized empty Git repository in C:/Users/shiftkey/src/test/repo/
$ ls
config  description  HEAD  hooks/  info/  objects/  refs/


I suspect there are some assumptions in the app around the presence of a .git directory within the folder that you provide. @ysabri<https://github.com/ysabri> can you tell me more about your use case here, and why you're avoiding initialising the repository in the typical way?ok il look out for That tomorroy for Now  its over  ty  and yes il try to hck out with new learn and more powerfûl computer and apps but i think i have not finish and im a weird guy so il’ i have to show all my code my porn and my Life the y can go hard  i  bélière in Karma! 🤪🤟👍

— lolll  yes first Time i fond c appli in my admin Microsoft top’ and ôta how i Wadjda discourer thaïs for thé rest i dont trust with un cause  for me you r thé crew theu want  fuck my lif e why and you want i sign the prof of how im wreck by you?

**Comment by batarmanrookie** [comment]
Envoyé de mon iPhone

Le 20 mars 2018 à 19:19, Brendan Forster <notifications@github.com<mailto:notifications@github.com>> a écrit :


So when I try and create a new repository in a directory that includes a bare repository, the app recognizes that the directory includes a git repository and asks me to add it instead.

To confirm I understand the issue, are you using git init --bare to create a Git repository and then adding that folder to the app?

$ cd test
$ mkdir repo
$ cd repo
$ git init --bare
Initialized empty Git repository in C:/Users/shiftkey/src/test/repo/
$ ls
config  description  HEAD  hooks/  info/  objects/  refs/


I suspect there are some assumptions in the app around the presence of a .git directory within the folder that you provide. @ysabri<https://github.com/ysabri> can you tell me more about your use case here, and why you're avoiding initialising the repository in the typical way?

Les LOL that  was not the en  plan but Now i learn better  Now 👹🤕🤣

**Comment by batarmanrookie** [comment]
Salut

Envoyé de mon iPhone

Le 20 mars 2018 à 19:52, Yazeed Sabri <notifications@github.com<mailto:notifications@github.com>> a écrit :


Sorry I approached the question weirdly. I have a bare repo with content in it already, meaning if I clone it I will be able to access a worktree with several commits in it. I tried to add it to the app and the add was successful, meaning the app recognizes it is a repository but not that it is a bare repository. My suggestion is for the app to able to recognize a bare repo, prevent me from adding it and maybe suggest that I perform a local clone of it.

I am agreeing with the fact that the app prevented me from creating a new repository where the bare repository resides. What I am not agreeing with is that the app let me add it. This stems from the fact that rev-prase --show-cdup does not convey information about the repository being bare or not, which might lead to the false assumption that a bare repository is addable to the app.

I know this a weird case, but it is not impossible, I am just pointing it out. I know how to setup the repository normally and use it, I was just trying it out on the app when I saw the app only uses --show-cdup option.l’looolll she s as l’ill  bitch! And its look like à ransomware Couz  like this alternions  sometime i have à choice against différent apps first us a mention top  fixes an horloge juste whem i play atound my Cmos with my l’Ed gputweak or a terre apps like My. Telechrgement and  other massive download of a bunch of girls or strange think ( my color are like à photo négative Noé and of i dont reliais my credit card to sign in this us what i Will receive me and my git hub follower its fukin cool 😎 a chanc é for me in a stoner Ida about that exempt if i can tale in réel with thé Guy starter this 😈🤡👬

**Comment by shiftkey** [comment]
If we're :thumbsup: with the underlying issue here (we should block adding a bare repository rather than considering it an uninitialized repository) then I'll port this to a fresh issue that someone can contribute the fix for:

> The app rejects adding it since it does not have a working tree or offers to do a local clone.

There's a way to check if a folder is a bare Git repository:

```
$ git rev-parse --is-bare-repository 
```

We should do this within the `AddExistingRepository` dialog and report a specific error to the dialog if we encounter it, to block them from adding it to the app.

**Comment by shiftkey** [comment]
For future reference, if we find ourselves supporting bare repositories in the app in the future:

> However, I have GitX installed, and it makes bare `foo.git` repos into a package, so we might need to add that back in if we start supporting bare repositories.

This will affect dragging-and-dropping the bare repository onto the Desktop tray icon in macOS, see #5207 for the current implementation of this support.

**Comment by j-f1** [comment]
We might want to make bare git repos into bundles ourselves once we add support. (a requirement for that is to have an icon for the bundle)

**Comment by Daniel-McCarthy** [comment]
I would be interested in attempting to implement this functionality if it is not of immediate urgency.

I have already started looking into it. I have added a function that uses the `git rev-parse --is-bare-repository ` command to detect if the repository is bare. However, the part I need some direction on is reporting the error to the user.

From what I can tell this dialog does not display errors like the Publish Repository dialog does. Should this be implemented first, or should we display the error as a pop-up of sorts?

**Comment by shiftkey** [comment]
@Daniel-McCarthy 

> From what I can tell this dialog does not display errors like the Publish Repository dialog does.

If you try and add a folder that doesn't yet have a `.git` directory you'll see this warning:

<img width="416" alt="screen shot 2018-08-18 at 12 28 01 pm" src="https://user-images.githubusercontent.com/359239/44300729-447ab080-a2e2-11e8-96f1-a6d596c533de.png">

You'll note that this blocks the user from adding the folder, and clicking the `create a repository` link will send them to a different dialog.

> Should this be implemented first, or should we display the error as a pop-up of sorts?

For the moment let's leverage those styles, and once we've got a review in place we can try and find the other flows that need to be updated to handle this.

**Comment by Daniel-McCarthy** [comment]
Wonderful! Thanks again for your time. I will implement this in the suggested style and submit a pull request once it's ready.

**Comment by shiftkey** [comment]
This has been fixed in #5434 and will be available in the next update
